### ISSUE TITLE
Deprecated: You are retrieving the service locator from within the class...

### ISSUE THREAD

**Issue by kachaj7** [zendframework/zend-mvc#129]
> public function getServiceLocator()
> {
>         trigger_error(sprintf(
>             'You are retrieving the service locator from within the class %s. Please be aware that '
>             . 'ServiceLocatorAwareInterface is deprecated and will be removed in version 3.0, along '
>             . 'with the ServiceLocatorAwareInitializer. You will need to update your class to accept '
>             . 'all dependencies at creation, either via constructor arguments or setters, and use '
>             . 'a factory to perform the injections.',
>             get_class($this)
>         ), E_USER_DEPRECATED);
>     return $this->serviceLocator;
> }

This code in production, are you kidding me? I'm a developer and what changes will be made in next Version I will read in documentation from ZF3. Do you know @DEPRECATED??? I see the problem is for weeks in the Code. I think I must change the framework, if you are making such thinks and discussing that you will learn people to program well...

If I hide the errors, I don't see other thinks, so this is bad idea to inform people about future disabled functionality.

**Comment by weierophinney** [comment]
First: please watch your tone. You are using a _free_ framework, paying _nothing_ for support, and using language that is frankly rude and aggressive towards the many volunteers that work on the project. I strongly urge you to read the [code of conduct](https://github.com/zendframework/zend-mvc/blob/master/CONDUCT.md) before you comment or raise issues again, specifically points 7 and 8.

Next, I'll provide the view point of the project, based on 10+ years of working on the project.

`@deprecated` annotations are great. But nobody sees them. They're viewed even less than the CHANGELOG.md file, which points these things out in detail in a section titled, appropriately, "Deprecated". If we _only_ use annotations and/or the changelog, we get fallout once we release a new version that removes those things: "But I never saw the deprecation notices!" is the common complaint, and no amount of pointing to them helps the situation. Instead we get people jumping ship because we changed something, _even though we've been messaging that change in advance._

What we have discovered over the years is that if you want to signal deprecation, the best way to do it is _through the code itself_. The `E_USER_DEPRECATED` error constant was designed and released by the fine folks at php.net _precisely for these situations._ It allows a project to signal _via code_ features that should no longer be relied upon, in order to allow users to _prepare their code in advance_ of removal.

I'd much rather our users _prepare now_, particularly when there are documented paths for making your code work _in both versions_. If we only signal that through annotations and/or the changelog, we'll hit single digit percentages of our users. Doing so through a formal deprecation notice in the code means that users have to either choose to ignore the errors (via disabling deprecation notices in error reporting), _or update their code to eliminate the notice from triggering._ In either situation, we can have relative certainty that _our users know about the upcoming change._

tl;dr: the deprecation notice is a professional and responsible way to signal change to end-users and allow them to make their code forwards compatible with future versions.

**Comment by kachaj7** [comment]
Sorry for the tone, but the programming way that is made from last time is not possible to understand for me. 
What I mean: 
1 If changes are made in the same version I understand that if I update I don't need care about working right? Last update from ca. 2.3 to ca. 2.5 Version. There was restructuring of the Project which was destroying the compability to previous versions. I think this should't happen and that was the reason why I was angry (but I shouldn't) was this.

2.Warning. I was reading other issues in which I couldn't write anything so I opened this. If you are using such a way to show changes, for me this isn't good way because most of the programmers will disable the error, which has influance on future warnings. (so only opinion 7)
This warning was destroying everything page and tests...

**Comment by weierophinney** [comment]
> There was restructuring of the Project which was destroying the compability to previous versions.

Can you point to specific issues, please? The only complaint I've seen over and over again with the 2.5 releases was the updated minimum PHP version (bumping from 5.3 to 5.5). This is something we've done _many_ times over the years, and is largely not seen as a requirement for a new _major_ version of the project. Users can opt-in to the new minor release, or not. (More on that below.)

Assuming that a new minor version will _always_ "just work" is a fallacy. While we try very hard to stick to semantic versioning, there are specific changes introduced (typically in minor versions) that are impossible to reconcile:
- security patches. Some security fixes, out of necessity, need to break backwards compatibility.
- deprecations. Issuing deprecations only on major versions means a project must support them for the entire lifecycle of that major version. This is untenable, and _nobody_ follows that. (php.net _regularly_ deprecates functionality during minor versions.) Deprecated functionality is then removed on the next _major_ version.
- dependency updates. New features may require new dependencies. Maintenance overhead may require updates in dependency versions. These are typically performed in minor releases, but bugfixes may result in these occurring in maintenance releases. (These are not technically changes in code... and yet, for some reason, any change in dependencies often results in upset users.)

All of this leads to an important point: don't just blindly run `composer update`.

Commit your `composer.lock`, and run tests after updating. If there is breakage, and you do not have the time or resources to deal with it _right now_, revert to your previous `composer.lock`, re-install, and keep working.

Better yet: _you should use the `~` operator to pin to releases in the same minor version_. As an example, if you want to stick with the 2.4 series, specify `~2.4.0` (or a higher maintenance patch version). This ensures compatibility as you do updates, as maintenance versions very, very rarely break BC. You then have to deliberately update your project requirements when you are ready to update to newer minor or major versions of a given library. This approach _puts you in control of the update cycle_.

> If you are using such a way to show changes, _for me_ this isn't good way because most of the programmers will disable the error, which has influance on future warnings.

(emphasis mine)

You are _assuming_ that "most programmers" will disable the error, but without any data to back that assertion up. _We_ are assuming that this is just one possibility.

There are three ways to deal with this style of deprecation:
- Stick your head in the sand and just disable error reporting of `E_USER_DEPRECATED`. This is, as you note, not ideal, and, honestly, a sign that you need some senior developers on your team that understand the implications.
- Update your code to address the deprecated functionality. This is the ideal solution, as it makes your code forwards compatible with future updates.
- Rollback to a previous minor release, and pin to it in your `composer.json`/`composer.lock`, and defer the update to a later time when you can update your code properly.

Our hope is that one of the latter two will occur. However, if we deprecate quietly, via annotations and/or changelog, we _know_ from past experience that developers _will do nothing_, and instead defer the complaints until the next major version is released. We'd prefer to be noisy about the upcoming changes, and guide developers towards updating their code _now_, so it works both now and in the future.
